src/Microsoft.OData.ModelBuilder/Vocabularies/CapabilitiesVocabularyExtensionMethods.cs

```diff
@@ -0,0 +1,500 @@
+﻿// Copyright (c) Microsoft Corporation.  All rights reserved.
+// Licensed under the MIT License.  See License.txt in the project root for license information.
+
+using System;
+using System.Collections.Generic;
+using System.Diagnostics.Contracts;
+using System.Linq;
+using Microsoft.OData.Edm;
+using Microsoft.OData.Edm.Csdl;
+using Microsoft.OData.Edm.Vocabularies;
+using Microsoft.OData.ModelBuilder.Vocabularies.Capabilities;
+
+namespace Microsoft.OData.ModelBuilder.Vocabularies
+{
+    /// <summary>
+    /// Contains extension methods for <see cref="EdmModel"/> to set the query capabilities vocabulary.
+    /// </summary>
+    internal static class CapabilitiesVocabularyExtensionMethods
+    {
+        private static readonly IEnumerable<IEdmStructuralProperty> EmptyStructuralProperties = Enumerable.Empty<IEdmStructuralProperty>();
+        private static readonly IEnumerable<IEdmNavigationProperty> EmptyNavigationProperties = Enumerable.Empty<IEdmNavigationProperty>();
+
+        private static IEdmEnumType _navigationType;
+
+        /// <summary>
+        /// Set Org.OData.Capabilities.V1.CountRestrictions to target.
+        /// </summary>
+        /// <param name="model">The model referenced to.</param>
+        /// <param name="target">The target entity set to set the inline annotation.</param>
+        /// <param name="isCountable">This entity set can be counted.</param>
+        /// <param name="nonCountableProperties">These collection properties do not allow /$count segments.</param>
+        /// <param name="nonCountableNavigationProperties">These navigation properties do not allow /$count segments.</param>
+        public static void SetCountRestrictionsAnnotation(this EdmModel model, IEdmEntitySet target, bool isCountable,
+            IEnumerable<IEdmProperty> nonCountableProperties,
+            IEnumerable<IEdmNavigationProperty> nonCountableNavigationProperties)
+        {
+            if (model == null)
+            {
```

Comment:
Nitpick. Looks like there's some inconsistency in how the null check is being done since 'Contract.Assert' is being used in some cases eg. Contract.Assert(model != null)